src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs

```diff
@@ -64,10 +64,13 @@ public override BoundNode VisitTryStatement(BoundTryStatement node)
                 // to execute the finally block before occurring.  Also, we do not handle branches
                 // *into* the finally block.
                 SetState(finallyState);
-                var tryAndCatchPending = SavePending();
                 var unsetInFinally = AllBitsSet();
                 VisitFinallyBlock(node.FinallyBlockOpt, ref unsetInFinally);
-                foreach (var pend in tryAndCatchPending.PendingBranches)
+                
+                // all the branches that are still pending must have been introduced in try/finally 
+                // with exception of YieldReturn, the branches logically go through finally
```

Comment:
This comment contradicts the comment, above, regarding branches from the finally block.
It looks like the comment on entry to this block does not correspond to what the code does.
I do not think the changes in this method are correct (or necessary), and would
regress the quality of recovery in error scenarios involving branches out of the finally block.

Reply:
The comment above seems strange indeed. I did not change it, but could not quite understand what it wants to tell.

The root cause of the issue is that merging of the finally state was done before pending branches introduced in try and catch had a chance to be resolved (in case they are completely internal to try/catch). We must resolve and unpend those branches when exiting these blocks. 
That would be the most reliable solution. - Since nothing can branch into try or catch, keeping the branches to labels in those blocks is not very helpful and may only result in bugs like this

---
In reply to: [151213791](https://github.com/dotnet/roslyn/pull/23172#discussion_r151213791) [](ancestors = 151213791)

Reply:
I will scale back the fix to the minimum, but we still do need changes in this method. The part where we go through `tryAndCatchPending.PendingBranches` without giving them a chance to resolve, if the branches terminate in those same blocks, is essentially what causes the bug.

---
In reply to: [151224702](https://github.com/dotnet/roslyn/pull/23172#discussion_r151224702) [](ancestors = 151224702,151213791)